src/db/awsES/locations-dao.js

```diff
@@ -0,0 +1,191 @@
+import _ from 'lodash';
+import aws from 'aws-sdk';
+import config from 'config';
+import connectionClass from 'http-aws-es';
+import elasticsearch from 'elasticsearch';
+import esb from 'elastic-builder';
+
+import { parseQuery } from 'utils/parse-query';
+
+const {
+  domain,
+  accessKeyId,
+  secretAccessKey,
+} = config.get('dataSources.awsES');
+
+/**
+ * Parses query parameters and generates an elasticsearch query body
+ *
+ * @param {object} queryParams Query paramaters from GET /locations request
+ * @returns {object} Elasticsearch query body
+ */
+const buildQueryBody = (queryParams) => {
+  const parsedParams = parseQuery(queryParams);
+  let q = esb.boolQuery();
+  if (parsedParams.search !== undefined) {
+    q = esb.multiMatchQuery([
+      'attributes.name',
+      'attributes.arcgisAbbreviation',
+      'attributes.bannerAbbreviation',
+    ], parsedParams.search);
+  }
+
+  if (parsedParams.name !== undefined) {
+    if (parsedParams.name.operator === 'fuzzy') {
+      q.must(esb.fuzzyQuery('attributes.name', parsedParams.name.value));
+    } else {
+      q.must(esb.termQuery('attributes.name.keyword', parsedParams.name));
+    }
+  }
+
+  if (parsedParams.type !== undefined) {
+    q.must(esb.matchQuery('attributes.type', parsedParams.type[0]));
+  }
+
+  if (parsedParams.hasGiRestroom !== undefined) {
+    if (parsedParams.hasGiRestroom) {
+      q.must(esb.rangeQuery('attributes.girCount').gt(0));
+    } else {
+      q.must(esb.matchQuery('attributes.girCount', 0));
+    }
+  }
+
+  if (parsedParams.adaParkingSpaceCount && parsedParams.adaParkingSpaceCount.operator === '>=') {
+    q.must(esb.rangeQuery('attributes.adaParkingSpaceCount')
+      .gte(parsedParams.adaParkingSpaceCount.value));
+  }
+
+  if (parsedParams.motorcycleParkingSpaceCount && parsedParams.motorcycleParkingSpaceCount.operator === '>=') {
+    q.must(esb.rangeQuery('attributes.motorcycleParkingSpaceCount')
+      .gte(parsedParams.motorcycleParkingSpaceCount.value));
+  }
+
+  if (parsedParams.evParkingSpaceCount && parsedParams.evParkingSpaceCount.operator === '>=') {
+    q.must(esb.rangeQuery('attributes.evParkingSpaceCount')
+      .gte(parsedParams.evParkingSpaceCount.value));
+  }
```

Comment:
Some attributes apparently seems could be combined together e.g. all spaceCount and all abbreviations parameters. Can you combine all similar and improve the codes?

Reply:
I'm not sure how you would simplify this. We need to have one if statement for each query parameter because the parameters can be used in any combination with each other.

Reply:
Take spaceCounts as example, I think you can combine these three to something like this:

```js
const spaceCounts = [
  'adaParkingSpaceCount',
  'motorcycleParkingSpaceCount',
  'evParkingSpaceCount',
];

_.forEach(spaceCounts, (spaceCount) => {
  if (parsedParams[spaceCount] && parsedParams[spaceCount].operator === '>=') {
    q.must(esb.rangeQuery(`attributes.${spaceCount}`)
      .gte(parsedParams[spaceCount].value));
});
```

Same logic for other parameters which share similar logic.